### PR TITLE
resolve meshio issues

### PR DESCRIPTION
**Pull request by banesullivan**
I can't read Dolfin meshes using PyVista's `meshio` reader from #472. These changes are a start to a fix as 2D coordinates are returned...

@keurfonluu, can you check this out with the following example? You should have push access to this branch, so please feel free to edit here. 

```py
import pyvista as pv
from pyvista import examples

# Downloads a sample Dolfin mesh from our data repo: https://github.com/pyvista/vtk-data
saved_file, _ = examples.downloads._download_file("dolfin_fine.xml")

pv.read_meshio(saved_file, file_format="dolfin-xml")
```

```
ERROR:root:Number of components is different...can't set data
UnstructuredGrid (0x7fac111847c8)
  N Cells:	5400
  N Points:	0
  X Bounds:	1.000e+299, -1.000e+299
  Y Bounds:	1.000e+299, -1.000e+299
  Z Bounds:	1.000e+299, -1.000e+299
  N Arrays:	0
```

----

Note: this Dolfin mesh is originally from [`vtkplotter`](https://github.com/marcomusy/vtkplotter) - they have a lot of great example mesh files in the `vtkplotter/data/` directory that we should borrow for testing `meshio` further

### CODE REVIEW

File: pyvista/utilities/fileio.py [hunk 1]
```diff
@@ -274,37 +272,35 @@ def read_meshio(filename, file_format = None):
...
-        grid.GetPointData().AddArray(data)
-
+    grid.point_arrays.update(mesh.point_data)
     # Set cell data
-    for k, v in cell_data.items():
-        data = vtk.util.numpy_support.numpy_to_vtk(numpy.array(v, numpy.float64))
-        data.SetName(k)
-        grid.GetCellData().AddArray(data)
+    grid.cell_arrays.update(cell_data)
+
```
Comment (banesullivan) [inline pyvista/utilities/fileio.py]: FYI, @keurfonluu - PyVista manages the data attributes as dictionaries just like `meshio`

File: pyvista/utilities/fileio.py [hunk 2]
```diff
@@ -274,37 +272,35 @@ def read_meshio(filename, file_format = None):
...
                 if kk in cell_data:
-                    cell_data[kk] = numpy.concatenate((cell_data[kk], vv))
+                    cell_data[kk] = np.concatenate((cell_data[kk], vv))
                 else:
                     cell_data[kk] = vv
 
     # Create pyvista.UnstructuredGrid object
+    points = mesh.points
+    if points.shape[1] == 2:
+        points = np.c_[points, np.zeros(len(points))]
```
Comment (banesullivan) [inline pyvista/utilities/fileio.py]: Okay - this appears to have been the needed fix!

**Comment by banesullivan** [comment]
Okay, all is working now!! 🎉 

Screen Shot 2019-12-09 at 11 55 10 PM

File: pyvista/utilities/fileio.py (hunk 2 above)
Comment (banesullivan) [reply pyvista/utilities/fileio.py]: @keurfonluu, should we handle a 1D case as well?

Comment (banesullivan) [reply pyvista/utilities/fileio.py]: this appears to have broken the existing meshio tests

**Comment by banesullivan** [comment]
It would probably be best to change that Dolfin example in the gallery to show casing how we can read number of different file formats that `meshio` supports and PyVista/VTK doesn't as well as outline how to use the `file_format` argument

File: pyvista/utilities/fileio.py (hunk 2 above)
Comment (keurfonluu) [reply pyvista/utilities/fileio.py]: Not this change. It seems to be due to the same type conversion error I had in the following change in ``point_arrays`` and ``cell_arrays``. See below.

File: pyvista/utilities/fileio.py (hunk 1 above)
Comment (keurfonluu) [reply pyvista/utilities/fileio.py]: ``meshio`` point and cell data should be converted to ``np.array(arr, np.float64)`` (or ``np.float32``). I guess there is some kind of conversion that is carried in ``numpy_to_vtk`` depending on the type of the values of the input array.

File: pyvista/utilities/fileio.py [hunk 3]
```diff
@@ -274,37 +272,35 @@ def read_meshio(filename, file_format = None):
...
+        np.array(points, np.float64),
     )
 
     # Set point data
-    for k, v in mesh.point_data.items():
-        data = vtk.util.numpy_support.numpy_to_vtk(numpy.array(v, numpy.float64))
-        data.SetName(k)
-        grid.GetPointData().AddArray(data)
-
+    grid.point_arrays.update(mesh.point_data)
```
Comment (keurfonluu) [inline pyvista/utilities/fileio.py]:
```suggestion
    grid.point_arrays.update({k: np.array(v, np.float64) for k, v in mesh.point_data.items()})
```

File: pyvista/utilities/fileio.py [hunk 4]
```diff
@@ -274,37 +272,35 @@ def read_meshio(filename, file_format = None):
...
+        cells.append(np.hstack((np.full((len(v), 1), numnodes), v)).ravel())
         cell_type += [vtk_type] * len(v)
         next_offset = offset[-1] + numnodes + 1
 
         # Extract cell data
         if k in mesh.cell_data.keys():
             for kk, vv in mesh.cell_data[k].items():
                 if kk in cell_data:
-                    cell_data[kk] = numpy.concatenate((cell_data[kk], vv))
+                    cell_data[kk] = np.concatenate((cell_data[kk], vv))
```
Comment (keurfonluu) [inline pyvista/utilities/fileio.py]:
```suggestion
                    cell_data[kk] = np.concatenate((cell_data[kk], np.array(vv, np.float64)))
```

File: pyvista/utilities/fileio.py [hunk 5]
```diff
@@ -274,37 +272,35 @@ def read_meshio(filename, file_format = None):
...
                 else:
                     cell_data[kk] = vv
```
Comment (keurfonluu) [inline pyvista/utilities/fileio.py]:
```suggestion
                    cell_data[kk] = np.array(vv, np.float64)
```

File: pyvista/utilities/fileio.py (hunk 2 above)
Comment (keurfonluu) [inline pyvista/utilities/fileio.py]:
```suggestion
        points = np.hstack((points, np.zeros((len(points),1))))
```

File: pyvista/utilities/fileio.py [hunk 6]
```diff
@@ -274,37 +272,35 @@ def read_meshio(filename, file_format = None):
...
 
         # Extract cell data
         if k in mesh.cell_data.keys():
             for kk, vv in mesh.cell_data[k].items():
                 if kk in cell_data:
-                    cell_data[kk] = numpy.concatenate((cell_data[kk], vv))
+                    cell_data[kk] = np.concatenate((cell_data[kk], np.array(vv, np.float64)))
                 else:
-                    cell_data[kk] = vv
+                     cell_data[kk] = np.array(vv, np.float64)
```
Comment (keurfonluu) [inline pyvista/utilities/fileio.py]:
```suggestion
                    cell_data[kk] = np.array(vv, np.float64)
```

**Comment by banesullivan** [comment]
Thanks for the updates @keurfonluu! Looks like this is good to go now - could you give this an approved review when you think its ready?

**Comment by banesullivan** [comment]
Actually, now tests are failing on Windows Python 3.5 when importing `meshio`.... huh...

**Comment by akaszynski** [comment]
Azure is reporting a failure on Python 3.5

```
C:\hostedtoolcache\windows\Python\3.5.4\x64\lib\site-packages\meshio\wkt\__init__.py:1: in <module>
    from ._wkt import read, write
     File "C:\hostedtoolcache\windows\Python\3.5.4\x64\lib\site-packages\meshio\wkt\_wkt.py", line 23
       tin_pattern = rf"TIN\s*\((?:\s*{triangle_pattern}\s*,?)*\s*\)"
                                                                    ^
   SyntaxError: invalid syntax
```

**Comment by banesullivan** [comment]
I really don’t know what that error is... 

Also, I’m not seeing a testing pipeline on windows for meshio... does it not exist? If not, we may need to rethink adding it as a hard dependency before the next release

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by banesullivan** [comment]
Saying this because that error appears to be a syntax issue within meshio

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by akaszynski** [comment]
Pipeline is https://dev.azure.com/pyvista/PyVista/_build/results?buildId=1051

Error indeed seems to be a Python 3.5 syntax incompatibility.  Since this is with an older version of Py3.5, we should have an xpass for this.  @keurfonluu, can you have this fixed in `meshio`?

Perhaps we should simply add xfail for any dependency related tests.  I think we should still support `meshio`, but not fail when there are errors that are out of our control/scope.

**Comment by banesullivan** [comment]
This just worries me if a user tries to install PyVista on 3.5 and then they can’t import because of this syntax issue when imported meshio

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by akaszynski** [comment]
Why not require it as a dependency but then import it as a nested import?

As a note, the error didn't occur on import but rather when that line was executed when reading the dolfin mesh.

**Comment by agramfort** [comment]
>
> Why not require it as a dependency but then import it as a nested import?
>
+1

**Comment by banesullivan** [comment]
The log I’m looking at on azure shows the error happening right as we `import meshio` in the `fileio.py` module. This means PyVista as a whole is unusable on Python 3.5 for windows. I think we have no choice but to make it a nested import (as @agramfort wanted in the first case 🙃 ) and request this be fixed upstream in `meshio`

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by keurfonluu** [comment]
> The log I’m looking at on azure shows the error happening right as we `import meshio` in the `fileio.py` module. This means PyVista as a whole is unusable on Python 3.5 for windows. I think we have no choice but to make it a nested import (as @agramfort wanted in the first case 🙃 ) and request this be fixed upstream in `meshio`
> 
> Sent with [GitHawk](http://githawk.com)

This should be fixed in the next version of ``meshio`` (fix has been committed but has not been deployed on PyPI yet). The issue was the presence of an f-string in a new file format. f-strings have only been introduced in Python 3.6.

**Comment by akaszynski** [comment]
Looks like the tests are still failing here:
https://github.com/pyvista/pyvista/blob/b3642eea7cef5eaf506095c8d510d887e62f5481/tests/test_meshio.py#L14-L18

@banesullivan, could you add an xfail?

**Comment by banesullivan** [comment]
@keurfonluu - can you all set up testing for all supported Python versions for `meshio`?

**Comment by keurfonluu** [comment]
> @keurfonluu - can you all set up testing for all supported Python versions for `meshio`?

Seems that multi-version test has been set up today.

**Comment by banesullivan** [comment]
@akaszynski, I'm hesitant to add an `xfail` there being that this is an important feature that we are now advertising (see #496)... I just added a `skipif` for Python 3.5 so that these IO functions are still tested on Python 3.6-7

**Comment by banesullivan** [comment]
As soon as the tests on this come back green, I'd like to merge as `master` is failing right now because of this syntax issue

An approved review would be greatly appreciated!

**Comment by banesullivan** [comment]
I'll clean up that `TO_TEST` thing later...

File: examples/00-load/read-dolfin.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+.. _FEniCS/Dolfin: https://fenicsproject.org
+"""
+import pyvista as pv
+from pyvista import examples
+
+###############################################################################
+# Let's download an example FEniCS/Dolfin mesh from our example data
+# repository. This will download an XML Dolfin mesh and save it to PyVista's
+# data directory.
+saved_file, _ = examples.downloads._download_file("dolfin_fine.xml")
```
Comment (agramfort) [inline examples/00-load/read-dolfin.py]: you should avoid exposing private functions in the public documentation.

Comment (banesullivan) [reply examples/00-load/read-dolfin.py]: Yeah, I agree... I did this to show that we are simply downloading a file and can pass that filename to the `pyvista.read()` method. We could use the `exampled.download_dolfin()` method, but it returns the loaded mesh without showing that it is as simple as passing the file name to `pyvista.read()`

We've been meaning to rewrite the whole downloads API for #210 so maybe we can make these utilities public when we rewrite it

**Pull request merged**